README.md
# LightTwinSVM

<h3>A simple, light-weight and fast implementation of standard Twin Support Vector Machine </h3>
<p align="center">
<a href="https://opensource.org/licenses/GPL-3.0"><img src="https://img.shields.io/badge/License-GPL%20v3-blue.svg" alt="License"></a>
<a href=""><img src="https://img.shields.io/pypi/pyversions/Django.svg" alt="Python Versions"></a>
<a href="https://github.com/mir-am/LightTwinSVM/releases"><img src="https://img.shields.io/github/release/mir-am/LightTwinSVM/all.svg" alt="latest release version"></a>
<a href='https://lighttwinsvm.readthedocs.io/en/latest/?badge=latest'><img src='https://readthedocs.org/projects/lighttwinsvm/badge/?version=latest' alt='Documentation Status' /></a>
<a href="https://travis-ci.org/mir-am/LightTwinSVM"><img src="https://api.travis-ci.org/mir-am/LightTwinSVM.svg?branch=master" alt="Travis-CI"></a>
<a href="https://ci.appveyor.com/project/mir-am/lighttwinsvm"><img src="https://ci.appveyor.com/api/projects/status/c625kl28aaqvuh9g?svg=true" alt="AppVeyor"></a>
</p>
<br />

1. [Introduction](#intro)
2. [Installation Guide](#installation-guide)
   - [Setup script](#setup-script-recommended)
   - [Building manually](#building-manually)
3. [User Guide](#user-guide)
   - [Usage example](#an-exmaple-of-using-command-line-interface)
   - [Tutorials](#tutorials)
   - [API documentation](#api-documentation)
4. [Dataset Format](#dataset-format)
5. [Support](#support)
6. [Citing LightTwinSVM](#citing-lighttwinsvm)
7. [Contributing](#contributing) 
8. [FAQ](#frequently-asked-questions)
9. [Donations](#donations)
10. [Numerical Experiments](#numerical-experiments) 

## Intro
LightTwinSVM is a simple and fast implementation of standard Twin Support Vector Machine. It is licensed under the terms of GNU GPL v3. Anyone who is interested in machine learning and classification can use this program for their work/projects.
 
The main features of the program are the following:
- A **simple console program** for running TwinSVM classifier
- **Fast optimization algorithm:** The clipDCD algorithm was improved and is implemented in C++ for solving optimization problems of TwinSVM.
- **Linear**, **RBF** kernel and Rectangular are supported.
- Binary and **Multi-class classification** (One-vs-All & One-vs-One) are supported.
- The OVO estimator is **compatible with scikit-learn** tools such as GridSearchCV, cross_val_score, etc.
- The classifier can be evaluated using either **K-fold cross-validation** or **Training/Test** split.
- It supports **grid search** over C and gamma parameters.
- **CSV** and **LIBSVM** data files are supported.
- Detailed classification result will be saved in a spreadsheet file.

Twin Support Vector Machine classifier was proposed by: <br />
Khemchandani, R., & Chandra, S. (2007). Twin support vector machines for pattern classification. IEEE Transactions on pattern analysis and machine intelligence, 29(5), 905-910.

The clipDCD algorithm was proposed by: <br />
Peng, X., Chen, D., & Kong, L. (2014). A clipping dual coordinate descent algorithm for solving support vector machines. Knowledge-Based Systems, 71, 266-278.

## Installation Guide
Currently, supported operating systems are as follows. Choose your OS from list below for detailed install instructions.
- [Debian-based Linux systems](#linux--mac-os-x) (Ubuntu 14.04, Ubuntu 16.04, Ubuntu 17.10, Ubuntu 18.04 and Linux Mint 18)
- [RPM-based Linux systems](#linux--mac-os-x) (Fedora)
- [Mac OSX](#linux--mac-os-x)
- [Microsoft Windows](#windows)

### Dependencies
First of all, [Python](https://www.python.org/) 3.5 interpreter or newer is required. Python 3 is usually installed by default on most Linux distributions.
In order to build and run the program, the following Python packages are needed:
- [NumPy](https://www.numpy.org)
- [SciPy](https://www.scipy.org/)
- [Scikit-learn](http://scikit-learn.org/stable/index.html)
- [Pandas](https://pandas.pydata.org/)
- [Pybind11](https://pybind11.readthedocs.io/en/stable/intro.html)
- [Cython](https://cython.org/)(To build C++ extension module on Windows.)
- [PyInstaller](https://www.pyinstaller.org/)(To generate a binary executable for Windows platform.)

In order to build C++ extension module(Optimizer), the following tools and libraries are required:
- [GNU C++ Compiler](https://gcc.gnu.org/) (For Linux systems)
- [Apple XCode](https://developer.apple.com/xcode/) (For OSX systems)
- [Visual Studio](https://visualstudio.microsoft.com/) (For Windows systems)
- [Armadillo C++ Linear Algebra Library](http://arma.sourceforge.net/)
- [LAPACK](http://www.netlib.org/lapack/) and [BLAS](http://www.netlib.org/blas/) Library


### Setup script (Recommended)
### Linux & Mac OS X
**A shell script is created to help users download required dependencies and install program automatically.** However, make sure that [Git](https://git-scm.com/) and GNU C++ compiler is installed on your system.

**A note for MacOS users:** Make sure that [Apple XCode](https://developer.apple.com/xcode/) is installed on your system.

To install the program, open a terminal and execute the following commands:
```
git clone https://github.com/mir-am/LightTwinSVM.git
cd LightTwinSVM && ./setup.sh
```
If the installation was successful, you'd be asked to delete temporary directory for installation. You can also run unit tests to check functionalities of the program. Finally, a Linux shell "ltsvm.sh" is created to run the program.
After the successful installation, LightTwinSVM program should look like this in terminal: <br />
![alt text](https://raw.githubusercontent.com/mir-am/LightTwinSVM/misc/img/LightTwinSVM.png)

### Windows
First, download Git program from [here](https://git-scm.com/) if it's not installed on your system. Also, [**Visual Studio 2015**](https://visualstudio.microsoft.com/) or newer should be installed so that C++ extension module can be compiled. Before proceeding further, make sure that all the required Python packages are installed. Dependencies are listed [here](#dependencies). 

**A note for Windows users:** If this is the first time that you will run a PowerShell script, then you need to make sure that the ExecutionPolicy is set on your system. Otherwise, you cannot run the setup script on Windows. Please check out this [answer](https://stackoverflow.com/a/4038991) on Stack Overflow that helps you set the ExecutionPolicy.

To install the program on Windows, open a PowerShell terminal and run the following commands:
```
git clone https://github.com/mir-am/LightTwinSVM.git
cd LightTwinSVM && .\win-setup.ps1
```
When the installation is finished, a batch file "ltsvm.bat" will be created to run the program.

### Building manually
It is highly recommended to install the LightTwinSVM program automatically using the setup script. If for some reasons you still want to build the program manually, a step-by-step guide is provided [here](https://github.com/mir-am/LightTwinSVM/wiki/Building-the-LightTwinSVM-manually-on-Linux-and-OSX-systems) for Linux and OSX systems.

## User Guide
### An example of using command line interface
LightTwinSVM is a simple console application. It has 4 steps for doing classification. Each step is explained below: <br />
**Step 1:** Choose your dataset by pressing Enter key. A file dialog window will be shown to help you find and select your dataset. CSV and LIBSVM files are supported. It is highly recommended to normalize your dataset. <br />
![alt text](https://github.com/mir-am/LightTwinSVM/blob/misc/img/LightTwinSVM-dataset.png)<br />
**Step 2:** Choose a kernel function among Linear, Gaussian (RBF) and Rectangular. RBF kernel often produces better classification result but takes more time. However if you want to use non-linear kernel and your dataset is large, then consider choosing Rectangular kernel.
<br />
```
Step 2/4: Choose a kernel function:(Just type the number. e.g 1)
1-Linear
2-RBF
3-RBF(Rectangular kernel)
-> 2
```
**Step 3:** To evaluate TwinSVM performance, You can either use [K-Fold cross validation](https://towardsdatascience.com/cross-validation-in-machine-learning-72924a69872f) or split your data into training and test sets. <br />
```
Step 3/4: Choose a test methodology:(Just type the number. e.g 1)
1-K-fold cross validation
2-Train/test split
-> 1
Determine number of folds for cross validation: (e.g. 5)
-> 5
```
**Step 4:** You need to determine the range of C penalty parameter and gamma (If RBF kernel selected.) for exhaustive grid search. <br /> 
An example:
```
Step 4/4:Type the range of C penalty parameter for grid search:
(Two integer numbers separated by space. e.g. -> -5 5
-> -4 4
```
After completing the above steps, the exhaustive search will be started. When the search process is completed, a detailed classification result will be saved in a spreadsheet file. In this file, all the common evalaution metrics(e.g Accuracy, Recall, Precision and F1) are provided.<br />
A instance of spreadsheet file containing classification result can be seen [here](https://github.com/mir-am/LightTwinSVM/blob/misc/TSVM_RBF_5-F-CV_pima-indian_2018-05-23%2013:21.csv).
 
### Tutorials
LightTwinSVM can be imported as a Python package in your project. Currently, a Jupyter notebook is avaliable [here](https://github.com/mir-am/LightTwinSVM/tree/master/docs/notebooks), which is "A Step-by-Step Guide on How to Use Multi-class TwinSVM".

To run the notebooks, make sure that Jupyter is installed on your system. If not, use the following command to install it:
```
pip3 install jupyter
```
For more details, check out [Jupyter documentation](https://jupyter.readthedocs.io/en/latest/index.html).

### API documentation
Aside from the program's command line interface, you may want to use the LightTwinSVM's Python package for your project. All you have to do is to copy-paste the "[ltsvm](https://github.com/mir-am/LightTwinSVM/tree/master/ltsvm)" folder (the **installed version**) into the root folder of your project. Next, you can import "ltsvm" package in a module of your interest.

You can read about the documentation of the LightTwinSVM's estimators and tools [here](https://lighttwinsvm.readthedocs.io/en/latest/index.html). 

## Dataset Format
- **LIBSVM** data files are supported. Note that the extension of this file should be '*.libsvm'.
- For **comma separated value (CSV)** file, make sure that your dataset is consistent with the following rules:
1. First row can be header names. (It's optional.)
2. First column should be labels of samples. Moreover, labels of positive and negative samples should be 1 and -1, respectively.
3. All the values in dataset except headernames should be numerical. Nominal values are not allowed. <br />
To help you prepare your dataset and test the program, three datasets are included [here](https://github.com/mir-am/LightTwinSVM/tree/master/dataset).

## Support
**Have a question about the software?**<br />
You can contact me via [email](mailto:mir-am@hotmail.com). Feedback and suggestions for improvements are welcome.<br />

**Have a problem with the software or found a bug?**<br />
To let me know and fix it, please open an issue [here](https://github.com/mir-am/LightTwinSVM/issues). <br />
To report a problem or bug, please provide the following information:<br />
1. Error messages<br />
2. Output of the program.<br />
3. Explain how to reproduce the problem if possible.

## Citing LightTwinSVM

[![status](http://joss.theoj.org/papers/a56708efcbf27c12352670b4647d92ca/status.svg)](http://joss.theoj.org/papers/a56708efcbf27c12352670b4647d92ca)

If you use the LightTwinSVM program in your research work, please cite the following paper:

- Mir et al., (2019). LightTwinSVM: A Simple and Fast Implementation of Standard Twin Support Vector Machine Classifier. Journal of Open Source Software, 4(35), 1252, https://doi.org/10.21105/joss.01252

BibTeX entry:

```
@article{ltsvm2019,
    title     = {LightTwinSVM: A Simple and Fast Implementation of Standard Twin Support Vector Machine Classifier},
    author    = {Mir, Amir M. and Nasiri, Jalal A.},
    journal   = {Journal of Open Source Software},
    volume    = {4},
    issue     = {35},
    pages     = {1252},
    year      = {2019},
    doi       = {10.21105/joss.01252},
    url       = {https://doi.org/10.21105/joss.01252}
}
```

## Contributing

CHANGELOG.md
# Changelog
All notable changes to LightTwinSVM program will be documented in this file. <br />
Main contributor: [Mir, A.](https://github.com/mir-am) (mir-am@hotmail.com) <br />
License: [GNU General Public License v3.0](https://github.com/mir-am/LightTwinSVM/blob/master/LICENSE.txt)<br />

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]
## [0.6.0] - 2019-03-31
### Added
- The JOSS paper was added to the repository.
- The [donation section](https://github.com/mir-am/LightTwinSVM#donations) and its button was added to README file.
- Added more debug info to the Bash script (setup.sh file).
- Added a [step-by-step guide](https://github.com/mir-am/LightTwinSVM/wiki/Building-the-LightTwinSVM-manually-on-Linux-and-OSX-systems) on how to build the LightTwinSVM manually.
- A [contributing guide](https://github.com/mir-am/LightTwinSVM/blob/master/CONTRIBUTING.md) was added to the project's wiki.

### Fixed
- BLAS libraries' path was fixed in PowerShell script.
- The optimizer can now be imported in the current session of PowerShell.
- Fixed import errors for building docs using Sphinx.
- Fixed compilation error of C++ extension module for OSX systems.
- Fixed spell error of the requirements file.

### Changed
- API documentation revised according to the [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) docstring guide.
- The results of Numerical Experiments in the README file was updated. 

## [0.5.0] - 2019-01-01
### Added
- The multi-class method of One-vs-All (OVA) was implemented.
- The multi-class method of One-vs-One (OVO) was implemented.
- The OVO estimator is scikit-learn compatible classifier.
- An option was added to command-line interface to choose between OVA and OVO schemes.
- LightTwinSVM is now Python package and its estimators and utilities can be imported in other Python projects.
- [API] get_param_names was added to esitmators' classes. It returns the hyper-parameters of the estimator.

### Changed
- Search space is now created using ParameterGrid of Scikit-learn.
- Validator class returns classification results as a Python dictionary.

### Fixed
- Wrong test size for Train/Test split was fixed.

## [0.4.0-alpha] - 2018-06-08
### Added
- Experimental support for Rectangular kenrnel.

### Changed
- eval_classfier.py, twinsvm.py and ui.py modules are re-written and improved.

## [0.3.0-alpha] - 2018-06-03
### Added
- Experimental support for Windows OS.

## [0.2.0-alpha] - 2018-05-30 
### Added
- [LIBSVM](https://www.csie.ntu.edu.tw/~cjlin/libsvmtools/datasets/) data is now supported.
- A customized progress bar for grid search

## [0.1.0-alpha] - 2018-05-24
### Added
- A simple console app created.
- Fast Optimization algorithm (ClippDCD optimizer)
- Both Linear and RBF kernel supported.
- K-fold cross validation supported.
- Training/Test split supported.
- Support for exhaustive grid search over C and gamma parameters.
- Detailed classification result in a spreadsheet file.


ltsvm/twinsvm.py
# -*- coding: utf-8 -*-

# LightTwinSVM Program - Simple and Fast
# Version: 0.6.0 - 2019-03-31
# Developer: Mir, A. (mir-am@hotmail.com)
# License: GNU General Public License v3.0

"""
Classes and functios are defined for training and testing TwinSVM classifier.

TwinSVM classifier generates two non-parallel hyperplanes.
For more info, refer to the original papar.
Khemchandani, R., & Chandra, S. (2007). Twin support vector machines for pattern classification. IEEE Transactions on pattern analysis and machine intelligence, 29(5), 905-910.

Motivated by the following paper, the multi-class TSVM is developed.
Tomar, D., & Agarwal, S. (2015). A comparison on multi-class classification methods based on least squares twin support vector machine. Knowledge-Based Systems, 81, 131-147.
"""


from sklearn.base import BaseEstimator, ClassifierMixin
from sklearn.utils.multiclass import check_classification_targets
from sklearn.utils.validation import check_X_y, check_is_fitted, check_array
from sklearn.utils import column_or_1d
import numpy as np

# ClipDCD optimizer is an extension module which is implemented in C++
from ltsvm.optimizer import clipdcd


class TSVM(BaseEstimator):
    
    """
    Twin Support Vector Machine for binary classification.
    
    Parameters
    ----------
    kernel : str, optional (default='linear')
        Type of the kernel function which is either 'linear' or 'RBF'.
    
    rect_kernel : float, optional (default=1.0)
        Percentage of training samples for Rectangular kernel.
        
    C1 : float, optional (default=1.0)
        Penalty parameter of first optimization problem.
        
    C2 : float, optional (default=1.0)
        Penalty parameter of second optimization problem.
        
    gamma : float, optional (default=1.0)
        Parameter of the RBF kernel function.
    
    Attributes
    ----------
    mat_C_t : array-like, shape = [n_samples, n_samples]
        A matrix that contains kernel values.
        
    cls_name : str
        Name of the classifier.
    
    w1 : array-like, shape=[n_features]
        Weight vector of class +1's hyperplane.
        
    b1 : float
        Bias of class +1's hyperplane.
        
    w2 : array-like, shape=[n_features]
        Weight vector of class -1's hyperplane.
    
    b2 : float
        Bias of class -1's hyperplane.
    
    """

    def __init__(self, kernel='linear', rect_kernel=1, C1=2**0, C2=2**0, \
                 gamma=2**0):

        self.C1 = C1
        self.C2 = C2
        self.gamma = gamma
        self.kernel = kernel
        self.rect_kernel = rect_kernel
        self.mat_C_t = None
        self.cls_name = 'TSVM'
        
        # Two hyperplanes attributes
        self.w1, self.b1, self.w2, self.b2 = None, None, None, None

    def get_params_names(self):
        
        """
        For retrieving the names of hyper-parameters of this classifier.
        
        Returns
        -------
        parameters : list of str, {['C1', 'C2'], ['C1', 'C2', 'gamma']}
            Returns the names of the hyperparameters which are same as
            the class' attributes.
        """
        
        return ['C1', 'C2'] if self.kernel == 'linear' else ['C1', 'C2', 'gamma']

    def fit(self, X_train, y_train):

        """
        It fits the binary TwinSVM model according to the given training data.
        
        Parameters
        ----------
        X_train : array-like, shape (n_samples, n_features) 
           Training feature vectors, where n_samples is the number of samples
           and n_features is the number of features. 
           
        y_train : array-like, shape(n_samples,)
            Target values or class labels.
           
        """

        # Matrix A or class 1 samples
        mat_A = X_train[y_train == 1]

        # Matrix B  or class -1 data 
        mat_B = X_train[y_train == -1]

        # Vectors of ones
        mat_e1 = np.ones((mat_A.shape[0], 1))
        mat_e2 = np.ones((mat_B.shape[0], 1))

        if self.kernel == 'linear':  # Linear kernel
            
            mat_H = np.column_stack((mat_A, mat_e1))
            mat_G = np.column_stack((mat_B, mat_e2))

        elif self.kernel == 'RBF': # Non-linear 

            # class 1 & class -1
            mat_C = np.row_stack((mat_A, mat_B))

            self.mat_C_t = np.transpose(mat_C)[:, :int(mat_C.shape[0] * self.rect_kernel)]

            mat_H = np.column_stack((rbf_kernel(mat_A, self.mat_C_t, self.gamma), mat_e1))

            mat_G = np.column_stack((rbf_kernel(mat_B, self.mat_C_t, self.gamma), mat_e2))


        mat_H_t = np.transpose(mat_H)
        mat_G_t = np.transpose(mat_G)

        # Compute inverses:
        # Regulariztion term used for ill-possible condition
        reg_term = 2 ** float(-7)

        mat_H_H = np.linalg.inv(np.dot(mat_H_t, mat_H) + (reg_term * np.identity(mat_H.shape[1])))
        mat_G_G = np.linalg.inv(np.dot(mat_G_t, mat_G) + (reg_term * np.identity(mat_G.shape[1])))

        # Wolfe dual problem of class 1
        mat_dual1 = np.dot(np.dot(mat_G, mat_H_H), mat_G_t)
        # Wolfe dual problem of class -1
        mat_dual2 = np.dot(np.dot(mat_H, mat_G_G), mat_H_t)

        # Obtaining Lagrange multipliers using ClipDCD optimizer
        alpha_d1 = np.array(clipdcd.clippDCD_optimizer(mat_dual1, self.C1)).reshape(mat_dual1.shape[0], 1)
        alpha_d2 = np.array(clipdcd.clippDCD_optimizer(mat_dual2, self.C2)).reshape(mat_dual2.shape[0], 1)

        # Obtain hyperplanes
        hyper_p_1 = -1 * np.dot(np.dot(mat_H_H, mat_G_t), alpha_d1)

        # Class 1
        self.w1 = hyper_p_1[:hyper_p_1.shape[0] - 1, :]
        self.b1 = hyper_p_1[-1, :]

        hyper_p_2 = np.dot(np.dot(mat_G_G, mat_H_t), alpha_d2)

        # Class -1
        self.w2 = hyper_p_2[:hyper_p_2.shape[0] - 1, :]
        self.b2 = hyper_p_2[-1, :]


    def predict(self, X_test):

        """
        Performs classification on samples in X using the TwinSVM model.
        
        Parameters
        ----------
        X_test : array-like, shape (n_samples, n_features)
            Feature vectors of test data.
                
        Returns
        -------
        output : array, shape (n_samples,)
            Predicted class lables of test data.
            
        """

        # Calculate prependicular distances for new data points 
        prepen_distance = np.zeros((X_test.shape[0], 2))

        kernel_f = {'linear': lambda i: X_test[i, :] , 'RBF': lambda i: rbf_kernel(X_test[i, :], \
                    self.mat_C_t, self.gamma)}

        for i in range(X_test.shape[0]):

            # Prependicular distance of data pint i from hyperplanes
            prepen_distance[i, 1] = np.abs(np.dot(kernel_f[self.kernel](i), self.w1) + self.b1)

            prepen_distance[i, 0] = np.abs(np.dot(kernel_f[self.kernel](i), self.w2) + self.b2)

        # Assign data points to class +1 or -1 based on distance from hyperplanes
        output = 2 * np.argmin(prepen_distance, axis=1) - 1

        return output


def rbf_kernel(x, y, u):

    """
    It transforms samples into higher dimension using Gaussian (RBF) kernel.
    
    Parameters
    ----------
    x, y : array-like, shape (n_features,)
        A feature vector or sample.
    
    u : float
        Parameter of the RBF kernel function.
        
    Returns
    -------
    float
        Value of kernel matrix for feature vector x and y.
    """

    return np.exp(-2 * u) * np.exp(2 * u * np.dot(x, y))


class HyperPlane:
    
    """
    Its object represents a hyperplane
    
    Attributes
    ----------
    w : array-like, shape (n_features,)
        Weight vector. If the RBF kernel is used, the shape will be (n_samples,)
        
    b : float
        Bias.
    """

    def __init__(self):

        self.w = None  # Coordinates of hyperplane
        self.b = None  # Bias term


class MCTSVM(BaseEstimator):

    """
    Multi-class Twin Support Vector Machine (One-vs-All Scheme)
    
    Parameters
    ----------
    kernel : str, optional (default='linear')
        Type of the kernel function which is either 'linear' or 'RBF'.
    
    C : float, optional (default=1.0)
        Penalty parameter.
        
    gamma : float, optional (default=1.0)
        Parameter of the RBF kernel function.
        
    Attributes
    ----------
    classifiers : dict
        Stores an intance of :class:`HyperPlane` class for each binary classifier.
        
    mat_D_t : list of array-like objects
        Stores kernel matrix for each binary classifier.
        
    cls_name : str
        Name of the classifier.
    """

    def __init__(self, kernel='linear', C=2**0, gamma=2**0):

        self.kernel = kernel
        self.C = C
        self.gamma = gamma
        self.classfiers = {}  # Classifiers
        self.mat_D_t = []  # For non-linear MCTSVM
        self.cls_name = 'TSVM_OVA'

    def get_params_names(self):
        
        """
        For retrieving the names of hyper-parameters of this classifier.
        
        Returns
        -------
        parameters : list of str, {['C'], ['C', 'gamma']}
            Returns the names of the hyperparameters which are same as
            the class' attributes.
        """
        
        return ['C'] if self.kernel == 'linear' else ['C', 'gamma']

    def fit(self, X_train, y_train):

        """
        It fits the OVA-TwinSVM model according to the given training data.
        
        Parameters
        ----------
        X_train : array-like, shape (n_samples, n_features) 
           Training feature vectors, where n_samples is the number of samples
           and n_features is the number of features. 
           
        y_train : array-like, shape(n_samples,)
            Target values or class labels.
        """

        num_classes = np.unique(y_train)

        # Construct K-binary classifiers
        for idx, i in enumerate(num_classes):

            # Samples of i-th class
            mat_X_i = X_train[y_train == i]

            # Samples of other classes
            mat_Y_i = X_train[y_train != i]

            # Vectors of ones
            mat_e1_i = np.ones((mat_X_i.shape[0], 1))
            mat_e2_i = np.ones((mat_Y_i.shape[0], 1))

            if self.kernel == 'linear':
                
                mat_A_i = np.column_stack((mat_X_i, mat_e1_i))
                mat_B_i = np.column_stack((mat_Y_i, mat_e2_i))

            elif self.kernel == 'RBF':

                mat_D = np.row_stack((mat_X_i, mat_Y_i))

                self.mat_D_t.append(np.transpose(mat_D))

                mat_A_i = np.column_stack((rbf_kernel(mat_X_i, self.mat_D_t[idx], self.gamma), mat_e1_i))
                mat_B_i = np.column_stack((rbf_kernel(mat_Y_i, self.mat_D_t[idx], self.gamma), mat_e2_i))

            mat_A_i_t = np.transpose(mat_A_i)
            mat_B_i_t = np.transpose(mat_B_i)

            # Compute inverses:
            # Regulariztion term used for ill-possible condition
            reg_term = 2 ** float(-7)
    
            mat_A_A = np.linalg.inv(np.dot(mat_A_i_t, mat_A_i) + (reg_term * np.identity(mat_A_i.shape[1])))
    
            # Dual problem of i-th class
            mat_dual_i = np.dot(np.dot(mat_B_i, mat_A_A), mat_B_i_t)
    
            # Obtaining Lagrange multipliers using ClippDCD optimizer
            alpha_i = np.array(clipdcd.clippDCD_optimizer(mat_dual_i, self.C)).reshape(mat_dual_i.shape[0], 1)
    
            hyperplane_i = np.dot(np.dot(mat_A_A, mat_B_i_t), alpha_i)
    
            hyper_p_inst = HyperPlane()
            hyper_p_inst.w = hyperplane_i[:hyperplane_i.shape[0] - 1, :]
            hyper_p_inst.b = hyperplane_i[-1, :]
    
            self.classfiers[i] = hyper_p_inst


    def predict(self, X_test):

        """
        Performs classification on samples in X using the OVA-TwinSVM model.
        
        Parameters
        ----------
        X_test : array-like, shape (n_samples, n_features)
            Feature vectors of test data.
                
        Returns
        -------
        output : array, shape (n_samples,)
            Predicted class lables of test data.
        """

        # Perpendicular distance from each hyperplane
        prepen_dist = np.zeros((X_test.shape[0], len(self.classfiers.keys())))

        kernel_f = {'linear': lambda i, j: X_test[i, :] , 'RBF': lambda i, j: rbf_kernel(X_test[i, :], \
                    self.mat_D_t[j], self.gamma)}

        for i in range(X_test.shape[0]):

            for idx, j in enumerate(self.classfiers.keys()):

                prepen_dist[i, idx] = np.abs(np.dot(kernel_f[self.kernel](i, idx), \
                           self.classfiers[j].w) + self.classfiers[j].b) / np.linalg.norm(self.classfiers[j].w)

        output = np.argmin(prepen_dist, axis=1) + 1

        return output


class OVO_TSVM(BaseEstimator, ClassifierMixin):

    """
    Multi Class Twin Support Vector Machine (One-vs-One Scheme)
    
    The :class:`OVO_TSVM` classifier is scikit-learn compatible, which means
    scikit-learn tools such as `cross_val_score <https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_val_score.html>`_ 
    and `GridSearchCV <https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html>`_
    can be used for an instance of :class:`OVO_TSVM`
    
    Parameters
    ----------
    kernel : str, optional (default='linear')
        Type of the kernel function which is either 'linear' or 'RBF'.
        
    C1 : float, optional (default=1.0)
        Penalty parameter of first optimization problem for each binary
        :class:`TSVM` classifier.
        
    C2 : float, optional (default=1.0)
        Penalty parameter of second optimization problem for each binary
        :class:`TSVM` classifier.
        
    gamma : float, optional (default=1.0)
        Parameter of the RBF kernel function.
        
    Attributes
    ----------
    cls_name : str
        Name of the classifier.
    
    bin_TSVM_models_ : list
        Stores intances of each binary :class:`TSVM` classifier.
    """    
    
    def __init__(self, kernel='linear', C1=1, C2=1, gamma=1):
               
        self.kernel = kernel
        self.C1 = C1
        self.C2 = C2
        self.gamma = gamma
        self.cls_name = 'TSVM_OVO'
        
    def get_params_names(self):
        
        """
        For retrieving the names of hyper-parameters of this classifier.
        
        Returns
        -------
        parameters : list of str, {['C1', 'C2'], ['C1', 'C2', 'gamma']}
            Returns the names of the hyperparameters which are same as
            the class' attributes.
        """
        
        return ['C1', 'C2'] if self.kernel == 'linear' else ['C1', 'C2', 'gamma']
         
    def _validate_targets(self, y):
        
        """
        Validates labels for training and testing classifier
        """
        y_ = column_or_1d(y, warn=True)
        check_classification_targets(y)
        self.classes_, y = np.unique(y_, return_inverse=True)
        
        return np.asarray(y, dtype=np.int)
    
    def _validate_for_predict(self, X):
        
        """
        Checks that the classifier is already trained and also test samples are
        valid
        """
        
        check_is_fitted(self, ['bin_TSVM_models_'])
        X = check_array(X, dtype=np.float64)
        
        n_samples, n_features = X.shape
        
        if n_features != self.shape_fit_[1]:
            
            raise ValueError("X.shape[1] = %d should be equal to %d," 
                             "the number of features of training samples" % 
                             (n_features, self.shape_fit_[1]))
        
        return X
    
    def fit(self, X, y):
        
        """
        It fits the OVO-TwinSVM model according to the given training data.
        
        Parameters
        ----------
        X : array-like, shape (n_samples, n_features) 
            Training feature vectors, where n_samples is the number of samples
            and n_features is the number of features.
           
        y : array-like, shape(n_samples,)
            Target values or class labels.
            
        Returns
        -------
        self : object
        """
        
        y = self._validate_targets(y)
        X, y = check_X_y(X, y, dtype=np.float64)
         
        # Allocate n(n-1)/2 binary TSVM classifiers
        self.bin_TSVM_models_ = ((self.classes_.size * (self.classes_.size - 1))
                               // 2 ) * [None]
        
        p = 0
        
        for i in range(self.classes_.size):
            
            for j in range(i + 1, self.classes_.size):
                
                #print("%d, %d" % (i, j))
                
                # Break multi-class problem into a binary problem
                sub_prob_X_i_j = X[(y == i) | (y == j)]
                sub_prob_y_i_j = y[(y == i) | (y == j)]
                
                #print(sub_prob_y_i_j)
                
                # For binary classification, labels must be {-1, +1}
                # i-th class -> +1 and j-th class -> -1
                sub_prob_y_i_j[sub_prob_y_i_j == j] = -1
                sub_prob_y_i_j[sub_prob_y_i_j == i] = 1
                
                self.bin_TSVM_models_[p] = TSVM(self.kernel, 1, self.C1, self.C2, \
                               self.gamma)
                
                self.bin_TSVM_models_[p].fit(sub_prob_X_i_j, sub_prob_y_i_j)
                
                p = p + 1
                
        self.shape_fit_ = X.shape
                
        return self
         
    def predict(self, X):
        
        """
        Performs classification on samples in X using the OVO-TwinSVM model.
        
        Parameters
        ----------
        X : array-like, shape (n_samples, n_features)
            Feature vectors of test data.
        
        Returns
        -------
        y_pred : array, shape (n_samples,)
            Predicted class lables of test data.
        """
        
        X = self._validate_for_predict(X)
        
        # Initialze votes
        votes = np.zeros((X.shape[0], self.classes_.size), dtype=np.int)
        
        # iterate over test samples
        for k in range(X.shape[0]):
            
            p = 0
        
            for i in range(self.classes_.size):
                
                for j in range(i + 1, self.classes_.size):
                    
                    y_pred = self.bin_TSVM_models_[p].predict(X[k, :].reshape(1, X.shape[1]))
                    
                    if y_pred == 1:
                        
                        votes[k, i] = votes[k, i] + 1
                        
                    else:
                        
                        votes[k, j] = votes[k, j] + 1
                        
                    p = p + 1
                        
        
         # Labels of test samples based max-win strategy
        max_votes = np.argmax(votes, axis=1)
            
        return self.classes_.take(np.asarray(max_votes, dtype=np.int))
                
        
if __name__ == '__main__':
    
#    from ltsvm.dataproc import read_data
#    from sklearn.metrics import accuracy_score
#    from sklearn.model_selection import train_test_split
    from sklearn.utils.estimator_checks import check_estimator
#    from sklearn.model_selection import cross_val_score, GridSearchCV
#    import time
#    
    check_estimator(OVO_TSVM)

    
#    train_data, labels, data_name = read_data('/home/mir/mir-projects/Mir-Repo/mc-data/wine.csv')
#    
#    X_train, X_test, y_train, y_test = train_test_split(train_data, labels,
#                                                        test_size=0.30, random_state=42)
#    
#    
##    param = {'C_1': [float(2**i) for i in range(-5, 6)],
##             'C_2': [float(2**i) for i in range(-5, 6)]}
#    
#    start_t = time.time()
##    
#    ovo_tsvm_model = MCTSVM()
#    ovo_tsvm_model.set_params(**{'C': 4, 'gamma': 0.1})
#    print(ovo_tsvm_model.get_params())
#    
#    #cv = cross_val_score(ovo_tsvm_model, train_data, labels, cv=10)
#    
##    result = GridSearchCV(ovo_tsvm_model, param, cv=10, n_jobs=4, refit=False, verbose=1)
##    result.fit(train_data, labels)
#    
#    print(X_train.shape)
##    
#    ovo_tsvm_model.fit(X_train, y_train)
#    
#    pred = ovo_tsvm_model.predict(X_test)
##    
#    print("Finished: %.2f ms" % ((time.time() - start_t) * 1000))
##    
#    print("Accuracy: %.2f" % (accuracy_score(y_test, pred) * 100))